client/txn.go

```diff
@@ -336,32 +346,56 @@ func (txn *Txn) exec(retryable func(txn *Txn) error) (err error) {
 }
 
 // send runs the specified calls synchronously in a single batch and
-// returns any errors.
+// returns any errors. If the transaction is read-only, a potential
+// EndTransaction call at the end is trimmed.
 func (txn *Txn) send(calls ...proto.Call) error {
 	if len(calls) == 0 {
 		return nil
 	}
-	txn.updateState(calls)
+	if err := txn.updateState(calls); err != nil {
+		return err
+	}
+
+	// If the transaction record indicates that the coordinator never wrote
+	// an intent (and the client doesn't have one lined up), then there's no
+	// need to send EndTransaction. If there is one anyways, cut it off.
+	if txn.haveEndTxn && !(txn.txn.Writing || txn.haveTxnWrite) {
```

Comment:
I see your comment. It does make sense, but leaves one dissatisfied. I dont have any suggestions at this stage, but let's think about this a bit more.